sip/sip-009-analysis-cost-assessment.md

```diff
@@ -0,0 +1,333 @@
+
+## Preamble
+
+Title: Clarity Parsing and Analysis Cost Assessment
+
+Author: Aaron Blankstein <aaron@blockstack.com>
+
+Status: Draft
+
+Type: Standard
+
+Created: 03/05/2020
+
+License: BSD 2-Clause
+
+# Abstract
+
+This document describes the measured costs and asymptotic costs
+assessed for parsing Clarity code into an abstract syntax tree (AST)
+and the static analysis of that Clarity code (type-checking and
+read-only enforcement). This will not specify the _constants_
+associated with those asymptotic cost functions. Those constants will
+necessarily be measured via benchmark harnesses and regression
+analyses.
+
+# Measurements for Execution Cost
+
+The cost of analyzing Clarity code is measured using the same 5 categories
+described in SIP-006 for the measurement of execution costs:
+
+1. Runtime cost: captures the number of cycles that a single
+   processor would require to process the Clarity block. This is a
+   _unitless_ metric, so it will not correspond directly to cycles,
+   but rather is meant to provide a basis for comparison between
+   different Clarity code blocks.
+2. Data write count: captures the number of independent writes
+   performed on the underlying data store (see SIP-004).
+3. Data read count: captures the number of independent reads
+   performed on the underlying data store.
+4. Data write length: the number of bytes written to the underlying
+   data store.
+5. Data read length: the number of bytes read from the underlying
+   data store.
+
+Importantly, these costs are used to set a _block limit_ for each
+block.  When it comes to selecting transactions for inclusion in a
+block, miners are free to make their own choices based on transaction
+fees, however, blocks may not exceed the _block limit_. If they do so,
+the block is considered invalid by the network --- none of the block's
+transactions will be materialized and the leader forfeits all rewards
+from the block.
+
+Costs for static analysis are assessed during the _type check_ pass.
+The read-only and trait-checking passes perform work which is strictly
+less than the work performed during type checking, and therefore, the
+cost assessment can safely fold any costs that would be incurred during
+those passes into the type checking pass.
+
+# Common Analysis Metrics and Costs
+
+## AST Parsing
+
+The Clarity parser has a runtime that is linear with respect to the Clarity
+program length.
+
+```
+a*X+b
+```
+
+where a and b are constants, and
+
+X := the program length in bytes
+
+## Dependency cycle detection
+
+Clarity performs cycle detection for intra-contract dependencies (e.g.,
+functions that depend on one another). This detection is linear in the
+number of dependency edges in the smart contract:
+
+```
+a*X+b
+```
+
+where a and b are constants, and
+X := the total number of dependency edges in the smart contract
+
+Dependency edges are created anytime a top-level definition refers 
+to another top-level definition.
```

Comment:
Is the number of edges equal to the number of times a contract must be looked up during parsing?  Or are there more edges created (or fewer)?

Reply:
No -- they're unrelated. The dependency edges are intra-contract not inter-contract. This is only important for the DefinitionSorter analysis, which does intra-contract cycle detection _and_ ensures that subsequent analysis passes check top-level definitions in an order that allows function definitions, etc. to be order agnostic.

Reply:
> Is it important that function definitions, etc. are order agnostic?

Yes -- we wanted to make sure function definitions were order agnostic. It's otherwise a fairly surprising feature in a language for order of definitions to matter. The cost of the check will be relatively low, since this is a pretty fast pass. If it turns out to be one of the more expensive passes in the smart contract checks, we can think about revisiting this decision.